### PR TITLE
Using Soundfile I/O operations + build_array functions from PySox

### PR DESCRIPTION
**Pull request by pseeth**
This PR achieves more milestones from #105. It does the following:

> - The current pysox master now can accept and return numpy arrays when doing tfm.build. By updating to this, we will be able to do everything in-memory, but will still have exec calls.
> - Once we use soundfile to load the audio files, we can get rid of sox.Combiner and should do all the summing of mixtures in-memory.

Gain adjustments of background and foreground events all use in-memory operations on the numpy arrays, which matches the gain adjustments made within sox (the code matches sox source code).

What Scaper does now:

```python
gain = self.ref_db + e.value['snr'] - fg_lufs
output_array = np.exp(gain * np.log(10) / 20) * output_array
```

What sox does under the hood:

```c
#ifndef M_LN10
#define M_LN10  2.30258509299404568402  /* natural log of 10 */
#endif

#define dB_to_linear(x) exp((x) * M_LN10 * 0.05)
```

### CODE REVIEW

File: scaper/core.py [hunk 1]
```diff
@@ -1698,38 +1698,42 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
-            tmpfiles = []
-            with _close_temp_files(tmpfiles):
-                isolated_events_audio_path = []
+            # List for storing all generated audio (one array for every event)
+            event_audio_arrays = []
+            isolated_events_audio_path = []
+            duration_in_samples = int(self.duration * self.sr)
 
-                role_counter = {'background': 0, 'foreground': 0}
+            role_counter = {'background': 0, 'foreground': 0}
```
Comment (justinsalamon) [inline scaper/core.py:1706]: what's the role of `role_counter`? doesn't seem to be used anywhere else?

File: scaper/core.py [hunk 2]
```diff
@@ -1698,38 +1698,42 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                        # correctly).
+                        tfm.convert(
+                            samplerate=self.sr,
+                            n_channels=self.n_channels,
+                            bitdepth=None
+                        )
+                        tfm.set_output_format(
+                            rate=self.sr,
+                            channels=self.n_channels
+                        )
```
Comment (justinsalamon) [inline scaper/core.py:1732]: Why is this needed? Is it because there's no output filename now?

File: scaper/core.py [hunk 3]
```diff
@@ -1698,38 +1698,42 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
                         # Then trim the duration of the background event
-                        cmb.trim(e.value['source_time'],
-                                 e.value['source_time'] +
-                                 e.value['event_duration'])
+                        tfm.trim(e.value['source_time'],
+                                    e.value['source_time'] +
+                                    e.value['event_duration'])
```
Comment (justinsalamon) [inline scaper/core.py:1736]: spurious tabs here

File: scaper/core.py [hunk 4]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
                             tmpfiles_internal.append(
                                 tempfile.NamedTemporaryFile(
                                     suffix='.wav', delete=False))
-                            # synthesize concatenated/trimmed background
-                            cmb.build(
-                                [e.value['source_file']] * ntiles,
-                                tmpfiles_internal[-1].name, 'concatenate')
+                            # read in background off disk
+                            event_audio_array, event_sample_rate = soundfile.read(
```
Comment (justinsalamon) [inline scaper/core.py:1746]: lets rename `event_audio_array` to `event_audio` (everywhere) to avoid being too close to `event_audio_arrays`

File: scaper/core.py [hunk 5]
```diff
@@ -1698,38 +1698,42 @@ def _generate_audio(self, audio_path, ann, reverb=None,
 
         with _set_temp_logging_level(temp_logging_level):
 
-            # Array for storing all tmp files (one for every event)
-            tmpfiles = []
-            with _close_temp_files(tmpfiles):
-                isolated_events_audio_path = []
+            # List for storing all generated audio (one array for every event)
+            event_audio_arrays = []
```
Comment (justinsalamon) [inline scaper/core.py:1702]: now that I'm reviewing this, I think `event_audio_list` might be a better name after all?

File: scaper/core.py [hunk 6]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                            # Write event_audio_array to disk so we can compute LUFS using ffmpeg
+                            soundfile.write(
+                                tmpfiles_internal[-1].name, event_audio_array.T, self.sr)
                             # NOW compute LUFS
                             bg_lufs = get_integrated_lufs(
                                 tmpfiles_internal[-1].name)
 
                             # Normalize background to reference DB.
                             gain = self.ref_db - bg_lufs
+                            gain = self.ref_db + e.value['snr'] - bg_lufs
```
Comment (justinsalamon) [inline scaper/core.py:1763]: the second line overwrites the first, so the first should go. BUT: why have we added `e.value['snr']` into this line?

File: scaper/core.py [hunk 7]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                            event_audio_array = np.exp(gain * np.log(10) / 20) * event_audio_array
```
Comment (justinsalamon) [inline scaper/core.py:1764]: is this numerically safe? (or, is this the safest way to achieve this?)

File: scaper/core.py [hunk 8]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
-
-                            # Prepare tmp file for output
-                            tmpfiles.append(
-                                tempfile.NamedTemporaryFile(
-                                    suffix='.wav', delete=False))
-
-                            tfm.build(tmpfiles_internal[-1].name,
-                                      tmpfiles[-1].name)
+                            event_audio_arrays.append(
+                                event_audio_array[:duration_in_samples])
```
Comment (justinsalamon) [inline scaper/core.py:1767]: just a reminder:
`event_audio_arrays` --> `event_audio_list`
`event_audio_array` --> `event_audio`

File: scaper/core.py [hunk 9]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                        # correctly).
+                        tfm.convert(
+                            samplerate=self.sr,
+                            n_channels=self.n_channels,
+                            bitdepth=None
+                        )
+                        tfm.set_output_format(
+                            rate=self.sr,
+                            channels=self.n_channels
+                        )
```
Comment (justinsalamon) [inline scaper/core.py:1784]: as before, just double checking we need this?

File: scaper/core.py [hunk 10]
```diff
@@ -1738,40 +1742,50 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
                         # Trim
                         tfm.trim(e.value['source_time'],
-                                 e.value['source_time'] +
-                                 e.value['event_duration'])
+                                    e.value['source_time'] +
+                                    e.value['event_duration'])
```
Comment (justinsalamon) [inline scaper/core.py:1788]: spurious tabs

File: scaper/core.py [hunk 11]
```diff
@@ -1785,7 +1799,7 @@ def _generate_audio(self, audio_path, ann, reverb=None,
                         # Apply very short fade in and out
                         # (avoid unnatural sound onsets/offsets)
                         tfm.fade(fade_in_len=self.fade_in_len,
-                                 fade_out_len=self.fade_out_len)
+                                    fade_out_len=self.fade_out_len)
```
Comment (justinsalamon) [inline scaper/core.py:1802]: remove tab

File: scaper/core.py [hunk 12]
```diff
@@ -1794,45 +1808,34 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                            
                             # synthesize edited foreground sound event
-                            tfm.build(e.value['source_file'],
-                                      tmpfiles_internal[-1].name)
-                            # if time stretched get actual new duration
-                            if e.value['time_stretch'] is not None:
-                                fg_stretched_duration = soundfile.info(
-                                    tmpfiles_internal[-1].name).duration
-
+                            event_audio_array, event_audio_rate = soundfile.read(
```
Comment (justinsalamon) [inline scaper/core.py:1813]: inconsistent with earlier `event_sample_rate`. Let's use `event_audio` and `event_sr` everywhere.

File: scaper/core.py [hunk 13]
```diff
@@ -1794,45 +1808,34 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
                             # NOW compute LUFS
                             fg_lufs = get_integrated_lufs(
                                 tmpfiles_internal[-1].name)
 
                             # Normalize to specified SNR with respect to
                             # background
-                            tfm = sox.Transformer()
                             gain = self.ref_db + e.value['snr'] - fg_lufs
-                            tfm.gain(gain_db=gain, normalize=False)
+                            event_audio_array = np.exp(gain * np.log(10) / 20) * event_audio_array
```
Comment (justinsalamon) [inline scaper/core.py:1828]: as before, is this numerically safe(est)?

File: scaper/core.py [hunk 14]
```diff
@@ -1794,45 +1808,34 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
-
-                            # Finally save result to a tmp file
-                            tmpfiles.append(
-                                tempfile.NamedTemporaryFile(
-                                    suffix='.wav', delete=False))
-                            tfm.build(tmpfiles_internal[-1].name,
-                                      tmpfiles[-1].name)
+                            prepad = int(self.sr * e.value['event_time'])
+                            postpad = max(0, duration_in_samples - (event_audio_array.shape[0] + prepad))
+                            event_audio_array = np.pad(event_audio_array, ((prepad, postpad)), mode='constant')
```
Comment (justinsalamon) [inline scaper/core.py:1834]: shall we also add `constant_values=0`, to defend against future numpy changes?

File: scaper/core.py (hunk 1 above)
Comment (pseeth) [reply scaper/core.py:1706]: `role_counter` is used when saving the isolated event audio. Here: https://github.com/justinsalamon/scaper/blob/ea72a7c7a67544b2df42395721f4572fe5bfa85e/scaper/core.py#L1844

File: scaper/core.py (hunk 2 above)
Comment (pseeth) [reply scaper/core.py:1732]: Yeah, we need to tell `sox` explicitly what the output format is so it can interpret what it's putting onto `stdout` correctly.

File: scaper/core.py (hunk 1 above)
Comment (pseeth) [reply scaper/core.py:1706]: This showed up in the diff for some mysterious reason.

File: scaper/core.py (hunk 3 above)
Comment (pseeth) [reply scaper/core.py]: Fixed.

File: scaper/core.py (hunk 4 above)
Comment (pseeth) [reply scaper/core.py]: Done.

File: scaper/core.py (hunk 5 above)
Comment (pseeth) [reply scaper/core.py]: Sounds good.

File: scaper/core.py (hunk 6 above)
Comment (pseeth) [reply scaper/core.py]: By mistake of course. Fixed it!

File: scaper/core.py (hunk 7 above)
Comment (pseeth) [reply scaper/core.py]: This is how it's done in the `sox` source code, so I went with it.

File: scaper/core.py [hunk 15]
```diff
@@ -1743,28 +1743,28 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                                tmpfiles_internal[-1].name, event_audio.T, self.sr)
                             # NOW compute LUFS
                             bg_lufs = get_integrated_lufs(
                                 tmpfiles_internal[-1].name)
 
                             # Normalize background to reference DB.
                             gain = self.ref_db - bg_lufs
-                            gain = self.ref_db + e.value['snr'] - bg_lufs
-                            event_audio_array = np.exp(gain * np.log(10) / 20) * event_audio_array
+                            gain = self.ref_db - bg_lufs
```
Comment (justinsalamon) [inline scaper/core.py]: duplicate, remove

**Comment by pseeth** [comment]
Great, removing the context manager and waiting for tests to pass. Then I'll squash and merge this into `issue-105` branch.
